src/legacy/index.ts

```diff
@@ -314,32 +312,42 @@ async function buildSubtree(
   const depInstanceIds = depGraph.getNodeDepsNodeIds(nodeId);
   if (!depInstanceIds || depInstanceIds.length === 0) {
     memoizationMap.set(nodeId, depTree);
-    return depTree;
+    return [depTree, false];
   }
 
   if (maybeDeduplicationSet) {
     if (maybeDeduplicationSet.has(nodeId)) {
       if (depInstanceIds.length > 0) {
         addLabel(depTree, 'pruned', 'true');
       }
-      return depTree;
+      return [depTree, false];
     }
     maybeDeduplicationSet.add(nodeId);
   }
 
+  let cyclePruned = false;
   for (const depInstId of depInstanceIds) {
+    if (ancestors.includes(depInstId)) {
+      cyclePruned = true;
+      addLabel(depTree, CYCLE_PRUNED_NODE_ID, 'true');
+      continue;
+    }
     // Deduplication of nodes occurs only within a scope of a top-level dependency.
     // Therefore, every top-level dep gets an independent set to track duplicates.
     if (isRoot && maybeDeduplicationSet !== false) {
       maybeDeduplicationSet = new Set();
     }
-    const subtree = await buildSubtree(
+    const [subtree, subtreeCyclePruned] = await buildSubtree(
       depGraph,
       depInstId,
       eventLoopSpinner,
       maybeDeduplicationSet,
+      ancestors?.concat([depInstId]),
```

Comment:
Surely `ancestors` can not be null, so we can drop the `?`. Also, we do not need to create throw-away array wrappers to pass to `concat`
```suggestion
      ancestors.concat(depInstId),
```